app/controllers/sitemap_controller.rb

```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+class SitemapController < ApplicationController
+  layout false
+  skip_before_action :preload_json, :check_xhr
+
+  def index
+    @sitemaps = Sitemap.where(enabled: true)
+
+    render :index
+  end
+
+  def page
+    index = params.require(:page)
+    sitemap = Sitemap.find_by(enabled: true, name: index.to_s)
+    raise Discourse::NotFound if sitemap.nil?
+
+    @output = Rails.cache.fetch("sitemap/#{index}/#{sitemap.max_page_size}", expires_in: 24.hours) do
```

Comment:
We need to be careful here since `index` or rather `params[:page]` is user input and this current implementation allows them to manipulate our cache key. I'm curious what is the use of  the param here in the cache key.

Reply:
We cache each page's results separately, so the index allows us to distinguish them. Note that using the user input is not risky here as:

- There's a constraint at the route level (`page: /[1-9][0-9]*/`)
- We validate that a sitemap for that page number exists in the DB

However, I think it's cleaner to use `sitemap#name` here instead of the user provider param.

Reply:
thank you for clarifying 👍 

> We validate that a sitemap for that page number exists in the DB
Where does this happen in the controller?

Reply:
Here:

```ruby
sitemap = Sitemap.find_by(enabled: true, name: index.to_s)
raise Discourse::NotFound if sitemap.nil?
```
For each existing page, there's a record in the DB where name is the page number.